Created an additional logger that logs to the systemd journal

jernst opened the pull request:
Usage: in `config.php`:

    'log_type' => 'systemd'

and all messages sent to the logger will go to the journal, mapped similarly to the `syslog` logger. This requires the `php-systemd` extension from here: https://github.com/systemd/php-systemd.

Advantage: a single place where to look what might be wrong with a running system.

jernst commented:
These failing tests seem unrelated to anything I did ...

go2sh commented:
You need to sign of your commits. ;-) See CONTRIBUTING.md

jernst committed:
Added a logger for systemd/journald
Added a unit test

Signed-off-by: Johannes Ernst <jernst@indiecomputing.com>

jernst commented:
@go2sh Done.

blizzz commented:
Works nicely! For NC 14 it should work with PHP 7.0, however it seems not to be the case yet: https://github.com/systemd/php-systemd/issues/5

Otherwise, do you know about (Debian) packages for CI?

jernst commented:
I developed this against PHP 7.2 on Arch / UBOS using this patch: https://github.com/systemd/php-systemd/pull/6 (I guess I should have mentioned that earlier).

Here is how I build the package: https://github.com/uboslinux/ubos-php/tree/master/php-systemd which comes from Arch AUR here: https://aur.archlinux.org/packages/php-systemd/

Sorry, I'm not current on how to build packages for debian, it's been a few years ...

jernst commented:
That php-systemd patch just got merged: https://github.com/systemd/php-systemd/pull/6

blizzz commented on lib/private/Log/Systemdlog.php:
```diff
@@ -0,0 +1,68 @@
+	public function write(string $app, $message, int $level) {
+		$journal_level = $this->levels[$level];
+		sd_journal_send('PRIORITY='.$journal_level,
+				'SYSLOG_IDENTIFIER=nextcloud',
```
instead of hardcoding nextcloud, it might be better to get the name from ThemingDefaults::getName() or have it configurable as syslog does.

jernst replied:
Yea, I was thinking about that, but then stopped myself from overcomplicating matters :-). If this turns out to be a problem in practice, we can simply extend it in the future.

blizzz replied:
Actually, we can reuse the `syslog_tag` config option, as it cannot be used in parallel anyway. Extra effort would only to rename it to something more generic, it requires a repair step though. Since you require IConfig in the constructor, but don't use it, it should either go, or just read the setting and store it in a class member. Then you need to just pass it as SYSLOG_IDENTIFIER. Does not really make things more complicated.

Another thing is that we should check whether `sd_journal_send` is available, to react decently if the module is not loaded. I'd check it in the constructor, too, and throw a HintException if it is not available.

blizzz commented:
I believe i taught the Debian 8 that we use in the test case to compile that module. I'll give a final and push tomorrow.

jernst committed:
Pull updates from https://github.com/nextcloud/server

Signed-off-by: Johannes Ernst <jernst@indiecomputing.com>

MorrisJobke commented on lib/private/Log/Systemdlog.php:
```diff
@@ -0,0 +1,68 @@
+ *
+ * @author Johannes Ernst <jernst@indiecomputing.com>
+ *
+ * @license AGPL-3.0
```
Would be nice if this is licensed under `GNU AGPL version 3 or any later version` - cc @schiessle

jernst replied:
Done.

blizzz commented:
@jernst [I added php-systemd](https://github.com/nextcloud/docker-ci/pull/96) to the corresponding CI container. To have effect, please apply this patch for CI to work:

```diff
diff --git a/.drone.yml b/.drone.yml
index 26a72c9b43..cf9b262abb 100644
--- a/.drone.yml
+++ b/.drone.yml
@@ -53,7 +53,7 @@ pipeline:
       matrix:
         TESTS: syntax-php7.1
   phan:
-    image: nextcloudci/php7.2:php7.2-11
+    image: nextcloudci/php7.2:php7.2-12
     commands:
       - composer install
       - composer require --dev "phan/phan:0.11.1"
@@ -170,7 +170,7 @@ pipeline:
         DB: NODB
         PHP: 7.1
   nodb-php7.2:
-    image: nextcloudci/php7.2:php7.2-11
+    image: nextcloudci/php7.2:php7.2-12
     commands:
       - NOCOVERAGE=true TEST_SELECTION=NODB ./autotest.sh sqlite
     when:
@@ -194,7 +194,7 @@ pipeline:
         DB: sqlite
         PHP: 7.1
   sqlite-php7.2:
-    image: nextcloudci/php7.2:php7.2-11
+    image: nextcloudci/php7.2:php7.2-12
     commands:
       - NOCOVERAGE=true TEST_SELECTION=DB ./autotest.sh sqlite
     when:
@@ -218,7 +218,7 @@ pipeline:
         DB: mysql
         PHP: 7.1
   mysql-php7.2:
-    image: nextcloudci/php7.2:php7.2-11
+    image: nextcloudci/php7.2:php7.2-12
     commands:
       - NOCOVERAGE=true TEST_SELECTION=DB ./autotest.sh mysql
     when:
@@ -292,7 +292,7 @@ pipeline:
         DB: mysqlmb4
         PHP: 7.1
   mysqlmb4-php7.2:
-    image: nextcloudci/php7.2:php7.2-11
+    image: nextcloudci/php7.2:php7.2-12
     commands:
       - NOCOVERAGE=true TEST_SELECTION=DB ./autotest.sh mysqlmb4
     when:
```

@jernst and may I ask for another favor? in config/sample.config.php update the comment about `'log_type' => 'file'` to reflect the other options and also state that the php-systemd module is needed for the systemd type. Also update the doc about `'syslog_tag'` if going for it as suggested ^

jernst commented:
Hang on, this was done in a hurry, will fix when I have more time.

blizzz commented on lib/private/Log/Syslog.php:
```diff
@@ -39,7 +39,7 @@ class Syslog implements IWriter {
 	];
 
 	public function __construct(IConfig $config) {
-		openlog($config->getSystemValue('syslog_tag', 'ownCloud'), LOG_PID | LOG_CONS, LOG_USER);
```
This change should be mentioned in https://github.com/nextcloud/server/issues/7827 after merge.

jernst committed:
Added Systemd.log documentation to config.sample.php
Changed name of default system (not systemd) logger from ownCloud to Nextcloud, to be consistent
Fixed license per https://github.com/nextcloud/server/pull/9760#discussion_r195026784
Pulled upstream updates

Signed-off-by: Johannes Ernst <jernst@indiecomputing.com>

jernst commented:
This should be better. Manually tested to 1) actually log to the journal, 2) `syslog_tag` works with journal, 3) `HintException` is thrown if the PHP systemd module is not active. Also signed commit, and call the extension `system` not `php-system` (that's the name of the package) in the comments, as the `php.ini` file needs `extension=systemd.so`.

blizzz commented:
@jernst This looks good! could you commit the patch in https://github.com/nextcloud/server/pull/9760#issuecomment-396888591 to make the tests pass?

jernst committed:
Applied patch for CI to work per https://github.com/nextcloud/server/pull/9760#issuecomment-396888591

Signed-off-by: Johannes Ernst <jernst@indiecomputing.com>

jernst commented:
This integration test failure seems unrelated to this PR?

blizzz commented:
It is nothing that i have seen failing before. Locally I also get different results. I restart the tests, if it reoccurs, could you rebase? if it still reoccurs, it could be related… though it appears strange.

blizzz commented:
@jernst So, locally I have RedisTest succeeding. Perhaps there was a glitch inbetween. Could you rebase to current master  and force-push?

blizzz commented:
I put #10048 in place as today is feature freeze. If that runs through, we may merge it today.

blizzz commented:
→ continue in #10048
